File: app/src/main/java/com/helloelliote/util/image/ImageUtil.java
package com.helloelliote.util.image;

import android.content.Context;
import android.database.Cursor;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.net.Uri;
import android.os.Environment;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;

import static android.graphics.Bitmap.createScaledBitmap;
import static android.graphics.BitmapFactory.decodeFile;
import static android.os.Environment.DIRECTORY_DCIM;
import static android.provider.MediaStore.MediaColumns.DATA;
import static android.provider.MediaStore.MediaColumns.DISPLAY_NAME;
import static java.util.Objects.requireNonNull;

public final class ImageUtil {

    private static final String[] PROJECTION_DATA = {DATA};
    private static final String[] PROJECTION_NAME = {DISPLAY_NAME};
    private static final String DATE_PATTERN = "yyyyMMdd_HHmmss_SSS";
    private static final String REGEX_IMAGE_EXT = "((\\.(?i)(jpg|jpeg|tif|tiff|webp|png|gif|bmp))$)";

    private static String uriToFilePath(@NonNull Context context, Uri uri) {
        Cursor cursor = context.getContentResolver().query(uri, PROJECTION_DATA, null, null, null);
        requireNonNull(cursor).moveToNext();
        String path = cursor.getString(cursor.getColumnIndex(DATA));
        cursor.close();
        return path;
    }

    @Nullable
    public static String uriToFileName(@NonNull Context context, @NonNull Uri uri) {
        if (!requireNonNull(uri.getScheme()).equals("file"))
            try (Cursor cursor = context.getContentResolver().query(uri, PROJECTION_NAME, null, null, null)) {
                if (cursor != null && cursor.moveToFirst()) {
                    return cursor.getString(cursor.getColumnIndex(DISPLAY_NAME));
                }
            }
        else return uri.getLastPathSegment();
        return null;
    }

    @NonNull
    public static File uriToFile(Context context, Uri uri) {
        return new File(uriToFilePath(context, uri));
    }

    public static File prepareFile() throws IOException {
        String timeStamp = new SimpleDateFormat(DATE_PATTERN, Locale.getDefault()).format(new Date());
        String imageFileName = String.format("IMG_%s_", timeStamp);
        File storageDir = new File(Environment.getExternalStoragePublicDirectory(DIRECTORY_DCIM), "Camera");
        return File.createTempFile(imageFileName, ".jpg", storageDir);
    }

    @SuppressWarnings("SameParameterValue")
    public static File subSample4x(@NonNull File file, final int maxResolution) {
        BitmapFactory.Options options = new BitmapFactory.Options();
        options.inSampleSize = 4;
        Bitmap sourceBitmap = decodeFile(file.getPath(), null);
        Bitmap copyBitmap = sourceBitmap.copy(sourceBitmap.getConfig(), true);
        float width = copyBitmap.getWidth();
        float height = copyBitmap.getHeight();
        String timeStamp = new SimpleDateFormat(DATE_PATTERN, Locale.getDefault()).format(new Date());
        File newFile = new File(file.getParent(), String.format("IMG_%s.jpg", timeStamp));
        try (FileOutputStream outputStream = new FileOutputStream(newFile)) {
            if (width < (float) maxResolution && height < (float) maxResolution) {
                copyBitmap.compress(Bitmap.CompressFormat.JPEG, 100, outputStream);
                return newFile;
            } else {
                float bitmapRatio = width / height;
                int newWidth = maxResolution;
                int newHeight = maxResolution;
                if (1.0f > bitmapRatio) {
                    newWidth = (int) ((float) maxResolution * bitmapRatio);
                } else {
                    newHeight = (int) ((float) maxResolution / bitmapRatio);
                }
                Bitmap resizeBitmap = createScaledBitmap(copyBitmap, newWidth, newHeight, true);
                resizeBitmap.compress(Bitmap.CompressFormat.JPEG, 100, outputStream);
                return newFile;
            }
        } catch (IOException e) {
            e.printStackTrace();
            return file;
        }
    }

    private static String createFileName(@NonNull String fileName) {
        String newFileName = fileName.replaceAll(REGEX_IMAGE_EXT, "_R.jpg");
        while (newFileName.indexOf(".") != newFileName.lastIndexOf(".")) {
            newFileName = newFileName.replaceFirst("(\\.(?i))", "");
        }
        newFileName = newFileName.replaceAll("(₩(?i))", "");
        return newFileName;
    }
}

File: app/src/main/java/com/helloelliote/util/image/ImageUtil.kt
package com.helloelliote.util.image

import android.content.Context
import android.graphics.Bitmap
import android.graphics.Bitmap.createScaledBitmap
import android.graphics.BitmapFactory
import android.graphics.BitmapFactory.decodeFile
import android.net.Uri
import android.os.Environment
import android.os.Environment.DIRECTORY_DCIM
import android.provider.MediaStore.MediaColumns.DATA
import android.provider.MediaStore.MediaColumns.DISPLAY_NAME
import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import java.text.SimpleDateFormat
import java.util.*

object ImageUtil {

    private val PROJECTION_DATA = arrayOf(DATA)
    private val PROJECTION_NAME = arrayOf(DISPLAY_NAME)
    private const val DATE_PATTERN = "yyyyMMdd_HHmmss_SSS"
    private const val REGEX_IMAGE_EXT = "((\\.(?i)(jpg|jpeg|tif|tiff|webp|png|gif|bmp))$)"

    private fun uriToFilePath(context: Context, uri: Uri): String {
        val cursor = context.contentResolver.query(uri, PROJECTION_DATA, null, null, null)
        var path = ""
        cursor?.let {
            it.moveToNext()
            path = it.getString(cursor.getColumnIndex(DATA))
            it.close()
        }
        cursor?.close()
        return path
    }

    fun uriToFile(context: Context, uri: Uri): File {
        return File(uriToFilePath(context, uri))
    }

    @Throws(IOException::class)
    fun prepareFile(): File {
        val timeStamp = SimpleDateFormat(DATE_PATTERN, Locale.getDefault()).format(Date())
        val imageFileName = "IMG_$timeStamp"
        val storageDir =
            File(Environment.getExternalStoragePublicDirectory(DIRECTORY_DCIM), "Camera")
        return File.createTempFile(imageFileName, ".jpg", storageDir)
    }

    fun subSample4x(file: File, maxResolution: Int): File {
        val options = BitmapFactory.Options()
        options.inSampleSize = 4
        val sourceBitmap = decodeFile(file.path, null)
        val copyBitmap = sourceBitmap.copy(sourceBitmap.config, true)
        val width = copyBitmap.width.toFloat()
        val height = copyBitmap.height.toFloat()
        val timeStamp = SimpleDateFormat(DATE_PATTERN, Locale.getDefault()).format(Date())
        val newFile = File(file.parent, "IMG_$timeStamp.jpg")
        try {
            FileOutputStream(newFile).use { outputStream ->
                if (width < maxResolution.toFloat() && height < maxResolution.toFloat()) {
                    copyBitmap.compress(Bitmap.CompressFormat.JPEG, 100, outputStream)
                    return newFile
                } else {
                    val bitmapRatio = width / height
                    var newWidth = maxResolution
                    var newHeight = maxResolution
                    if (1.0f > bitmapRatio) {
                        newWidth = (maxResolution.toFloat() * bitmapRatio).toInt()
                    } else {
                        newHeight = (maxResolution.toFloat() / bitmapRatio).toInt()
                    }
                    val resizeBitmap = createScaledBitmap(copyBitmap, newWidth, newHeight, true)
                    resizeBitmap.compress(Bitmap.CompressFormat.JPEG, 100, outputStream)
                    return newFile
                }
            }
        } catch (e: IOException) {
            e.printStackTrace()
            return file
        }
    }

    private fun createFileName(fileName: String): String {
        var newFileName = fileName.replace(REGEX_IMAGE_EXT.toRegex(), "_R.jpg")
        while (newFileName.indexOf(".") != newFileName.lastIndexOf(".")) {
            newFileName = newFileName.replaceFirst("(\\.(?i))".toRegex(), "")
        }
        newFileName = newFileName.replace("(₩(?i))".toRegex(), "")
        return newFileName
    }
}
